File: frontend/src/api/agent.ts
import { apiClient } from './client'
import type { Chapter } from './client'

export interface PlotQuestion {
  id: string
  type: 'must_ask' | 'optional' | 'auto_decidable'
  question: string
  reason: string
  options: string[]
}

export interface QuestionAnswer {
  question_id: string
  answer: string
}

export interface ConsistencyIssue {
  severity: 'high' | 'medium' | 'low'
  type: 'character' | 'timeline' | 'worldbuilding' | 'plot' | 'foreshadowing' | 'style'
  description: string
  suggestion: string
}

export interface ConsistencyReport {
  summary: string
  issues: ConsistencyIssue[]
}

export interface StartChapterResponse {
  session_id: string
  next_action: string
  plot_questions: PlotQuestion[]
  auto_decidable: string[]
}

export interface AnswerQuestionsResponse {
  next_action: string
  chapter_outline: string
}

export interface ApproveOutlineResponse {
  next_action: string
  chapter_outline: string
  draft: string
  polished_draft: string
  consistency_report: ConsistencyReport | null
}

export async function startChapter(payload: {
  project_id: number
  chapter_id: number
  user_goal: string
}): Promise<StartChapterResponse> {
  const { data } = await apiClient.post<StartChapterResponse>('/agent/start-chapter', payload)
  return data
}

export async function answerQuestions(payload: {
  session_id: string
  answers: QuestionAnswer[]
}): Promise<AnswerQuestionsResponse> {
  const { data } = await apiClient.post<AnswerQuestionsResponse>('/agent/answer-questions', payload)
  return data
}

export async function approveOutline(payload: {
  session_id: string
  approved: boolean
  revision_instruction: string
}): Promise<ApproveOutlineResponse> {
  const { data } = await apiClient.post<ApproveOutlineResponse>('/agent/approve-outline', payload)
  return data
}

export async function saveChapter(payload: {
  session_id: string
  use_polished: boolean
}): Promise<Chapter> {
  const { data } = await apiClient.post<Chapter>('/agent/save-chapter', payload)
  return data
}

File: backend/app/api/agent.py
from __future__ import annotations

from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app import crud, models, schemas
from app.agent.graph import run_draft_phase, run_outline_phase, run_question_phase
from app.agent.nodes import save_chapter_from_state
from app.api.chapters import get_chapter_or_404
from app.api.projects import get_project_or_404
from app.database import get_db
from app.rag.indexer import rebuild_project_index

router = APIRouter(prefix="/api/agent", tags=["agent"])


def get_session_or_404(session_id: str, db: Session) -> models.AgentSession:
    item = db.get(models.AgentSession, session_id)
    if not item:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Agent session not found")
    return item


def _save_session(db: Session, item: models.AgentSession, state: schemas.NovelAgentState) -> None:
    item.state_json = dict(state)
    item.next_action = str(state.get("next_action", item.next_action))
    db.commit()
    db.refresh(item)


@router.post("/start-chapter", response_model=schemas.StartChapterResponse)
def start_chapter(payload: schemas.StartChapterRequest, db: Session = Depends(get_db)):
    get_project_or_404(payload.project_id, db)

    if payload.chapter_id is None:
        chapter = crud.create_chapter(
            db,
            payload.project_id,
            schemas.ChapterCreate(
                chapter_number=crud.next_chapter_number(db, payload.project_id),
                title="未命名章节",
                goal=payload.user_goal,
                status="planning",
            ),
        )
    else:
        chapter = get_chapter_or_404(payload.project_id, payload.chapter_id, db)
        chapter.goal = payload.user_goal
        chapter.status = "planning"
        db.commit()
        db.refresh(chapter)

    rebuild_project_index(payload.project_id, db)
    initial_state: schemas.NovelAgentState = {
        "project_id": payload.project_id,
        "chapter_id": chapter.id,
        "user_goal": payload.user_goal,
        "messages": [],
    }
    state = run_question_phase(initial_state, db)
    session_id = uuid4().hex
    agent_session = models.AgentSession(
        id=session_id,
        project_id=payload.project_id,
        chapter_id=chapter.id,
        next_action=state.get("next_action", "answer_plot_questions"),
        state_json=dict(state),
    )
    db.add(agent_session)
    db.commit()

    return schemas.StartChapterResponse(
        session_id=session_id,
        next_action=agent_session.next_action,
        plot_questions=state.get("plot_questions", []),
        auto_decidable=state.get("auto_decidable", []),
    )


@router.post("/answer-questions", response_model=schemas.AnswerQuestionsResponse)
def answer_questions(payload: schemas.AnswerQuestionsRequest, db: Session = Depends(get_db)):
    agent_session = get_session_or_404(payload.session_id, db)
    state = dict(agent_session.state_json or {})
    state["user_answers"] = [answer.model_dump() for answer in payload.answers]
    state = run_outline_phase(state, db)
    _save_session(db, agent_session, state)
    return schemas.AnswerQuestionsResponse(
        next_action=state.get("next_action", "approve_outline"),
        chapter_outline=state.get("chapter_outline", ""),
    )


@router.post("/approve-outline", response_model=schemas.ApproveOutlineResponse)
def approve_outline(payload: schemas.ApproveOutlineRequest, db: Session = Depends(get_db)):
    agent_session = get_session_or_404(payload.session_id, db)
    state = dict(agent_session.state_json or {})

    if not payload.approved:
        state["outline_approved"] = False
        state["outline_revision_instruction"] = payload.revision_instruction
        state = run_outline_phase(state, db)
        _save_session(db, agent_session, state)
        return schemas.ApproveOutlineResponse(
            next_action=state.get("next_action", "approve_outline"),
            chapter_outline=state.get("chapter_outline", ""),
        )

    state["outline_approved"] = True
    state = run_draft_phase(state, db)
    _save_session(db, agent_session, state)
    return schemas.ApproveOutlineResponse(
        next_action=state.get("next_action", "review_draft"),
        draft=state.get("draft", ""),
        polished_draft=state.get("polished_draft", ""),
        consistency_report=state.get("consistency_report"),
    )


@router.post("/save-chapter", response_model=schemas.ChapterRead)
def save_chapter(payload: schemas.SaveChapterRequest, db: Session = Depends(get_db)):
    agent_session = get_session_or_404(payload.session_id, db)
    state = dict(agent_session.state_json or {})
    try:
        chapter = save_chapter_from_state(state, db, payload.use_polished)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    state["next_action"] = "completed"
    _save_session(db, agent_session, state)
    rebuild_project_index(agent_session.project_id, db)
    return chapter
